src/transactions/typeTag/typeTagParser.ts

```diff
@@ -0,0 +1,246 @@
+/* eslint-disable class-methods-use-this */
+/* eslint-disable max-classes-per-file */
+import {
+  StructTag,
+  TypeTag,
+  TypeTagAddress,
+  TypeTagBool,
+  TypeTagParserError,
+  TypeTagStruct,
+  TypeTagU128,
+  TypeTagU16,
+  TypeTagU256,
+  TypeTagU32,
+  TypeTagU64,
+  TypeTagU8,
+  TypeTagVector,
+} from "./typeTag";
+import { AccountAddress } from "../../core";
+import { Identifier } from "../instances";
+
+function isValidIdentifier(str: string) {
+  return !!str.match(/^[_a-zA-Z0-9]+$/);
+}
+
+function isValidWhitespaceCharacter(char: string) {
+  return !!char.match(/\s/);
+}
+
+function consumeWhitespace(tagStr: string, pos: number) {
+  let i = pos;
+  for (; i < tagStr.length; i += 1) {
+    const innerChar = tagStr[i];
+
+    if (!isValidWhitespaceCharacter(innerChar)) {
+      // If it's not colons, and it's an invalid character, we will stop here
+      break;
+    }
+  }
+  return i;
+}
+
+/**
+ * All types are made of a few parts they're either:
+ * 1. A simple type e.g. u8
+ * 2. A standalone struct e.g. 0x1::account::Account
+ * 3. A nested struct e.g. 0x1::coin::Coin<0x1234::coin::MyCoin>
+ *
+ * There are a few more special cases that need to be handled, however.
+ * 1. Multiple generics e.g 0x1::pair::Pair<u8, u16>
+ * 2. Spacing in the generics e.g. 0x1::pair::Pair< u8 , u16>
+ * 3. Nested generics of different depths e.g. 0x1::pair::Pair<0x1::coin::Coin<0x1234::coin::MyCoin>, u8>
+ * 4. Generics for types in ABIs are filled in with placeholders e.g T1, T2, T3
+ */
+export function parseTypeTag(typeStr: string) {
+  const saved: Array<{ savedExpectedTypes: number; savedStr: string; savedTypes: Array<TypeTag> }> = [];
+  // This represents the internal types for a type tag e.g. '0x1::coin::Coin<innerTypes>'
+  let innerTypes: Array<TypeTag> = [];
+  // This represents the current parsed types in a comma list e.g. 'u8, u8'
+  let curTypes: Array<TypeTag> = [];
+  // This represents the current character index
+  let cur = 0;
+  // This represents the current working string as a type or struct name
+  let currentStr = "";
+  let expectedTypes = 1;
+
+  // Iterate through each character, and handle the border conditions
+  while (cur < typeStr.length) {
+    const char = typeStr[cur];
+
+    if (char === "<") {
+      // Start of a type argument, push current state onto a stack
+      saved.push({
+        savedExpectedTypes: expectedTypes,
+        savedStr: currentStr,
+        savedTypes: curTypes,
+      });
+
+      // Clear current state
+      currentStr = "";
+      curTypes = [];
+      expectedTypes = 1;
+    } else if (char === ">") {
+      // Process last type, if there is no type string, then don't parse it
+      if (currentStr !== "") {
+        const newType = parseTypeTagInner(currentStr, innerTypes);
+        curTypes.push(newType);
+      }
+
+      // Pop off stack outer type, if there's nothing left, there were too many '>'
+      const savedPop = saved.pop();
+      if (savedPop === undefined) {
```

Comment:
do we also need to check for `null`?

Reply:
`pop` returns `undefined` or `value` as far as I could see.